core/network/impl/protocols/grandpa_protocol.cpp

```diff
@@ -114,80 +121,80 @@ namespace kagome::network {
+                   stream_res.error().message());
+        cb(stream_res.as_failure());
+        return;
+      }
+      auto &stream = stream_res.value();
 
-            SL_VERBOSE(self->log_,
-                       "Fully established outgoing {} stream with {}",
-                       self->protocol_,
-                       stream->remotePeerId().value());
-
-            // Send neighbor message first
-            auto own_peer_state =
-                self->peer_manager_->getPeerState(self->own_info_.id);
-            if (own_peer_state.has_value()) {
-              self->neighbor(GrandpaNeighborMessage{
-                  .round_number = own_peer_state->round_number.value_or(1),
-                  .voter_set_id = own_peer_state->set_id.value_or(0),
-                  .last_finalized = own_peer_state->last_finalized});
-            }
-
-            cb(std::move(stream));
-          };
-
-          self->writeHandshake(
-              std::move(stream), Direction::OUTGOING, std::move(cb2));
-        });
+      auto cb2 = [wp, stream, cb = std::move(cb)](outcome::result<void> res) {
+        auto self = wp.lock();
+        if (not self) {
+          cb(ProtocolError::GONE);
+          return;
+        }
+
+        if (not res.has_value()) {
+          SL_VERBOSE(self->log_,
+                     "Handshake failed on outgoing {} stream with {}: {}",
+                     self->protocol_,
+                     stream->remotePeerId().value(),
+                     res.error().message());
+          stream->reset();
+          cb(res.as_failure());
+          return;
+        }
+
+        res = self->stream_engine_->addOutgoing(stream, self);
+        if (not res.has_value()) {
+          SL_VERBOSE(self->log_,
+                     "Can't register outgoing {} stream with {}: {}",
+                     self->protocol_,
+                     stream->remotePeerId().value(),
+                     res.error().message());
+          stream->reset();
+          cb(res.as_failure());
+          return;
+        }
+
+        SL_VERBOSE(self->log_,
+                   "Fully established outgoing {} stream with {}",
+                   self->protocol_,
+                   stream->remotePeerId().value());
+
+        // Send neighbor message first
+        auto own_peer_state =
+            self->peer_manager_->getPeerState(self->own_info_.id);
+        if (own_peer_state.has_value()) {
+          self->neighbor(GrandpaNeighborMessage{
+              .round_number = own_peer_state->round_number.value_or(1),
+              .voter_set_id = own_peer_state->set_id.value_or(0),
+              .last_finalized = own_peer_state->last_finalized});
+        }
+
+        cb(std::move(stream));
+      };
+
+      self->writeHandshake(
+          std::move(stream), Direction::OUTGOING, std::move(cb2));
+    };
+
+    host_.newStream(peer_info.id, protocol_, stream_handler);
+    host_.newStream(peer_info.id, protocol_by_genesis_, stream_handler);
```

Comment:
Oh, no!
Please, try to establish genesis based stream first. And try old protocol name in case of failure.